### PR TITLE
Fix #5014: Wrong calculation of width for scrollable tables

### LINKED ISSUE #5014: Datatable: (scrollable): size column size calculation defect for percentage-width or no width given
Version: Primefaces 6.2.21

## 2) Expected behavior

when no width given, all columns should have same size and be distributed equally over the available scrollWidth

when width="30%", this should be 30% and not 30px!

## 3) Actual behavior

since commit https://github.com/primefaces/primefaces/commit/852741bf1376fc8e966441bffc9a82e80c601b15 to solve issue https://github.com/primefaces/primefaces/issues/4801

column width is NOT calculated correctly anymore, if NOT using box-sizing:border-box

- 30% is interpreted as 30px

- and when no width given, its distributed over the window-width instead of the given scrollWidth

- and columns on the left are small and further right columns get wider

## 4) Steps to reproduce

any datatable with scrollable=true

## 5) Sample XHTML

Primefaces 6.2.21 and 

`<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html lang="en" xmlns="http://www.w3.org/1999/xhtml"
      xmlns:c="http://java.sun.com/jsp/jstl/core"
      xmlns:ui="http://java.sun.com/jsf/facelets"
      xmlns:h="http://java.sun.com/jsf/html"
      xmlns:f="http://java.sun.com/jsf/core"
      xmlns:p="http://primefaces.org/ui" >

	<f:view contentType="text/html; charset=utf-8" >

		<h:head></h:head>

		<h:body>

		   <h3>Vertical</h3>
    <p:dataTable var="car" value="#{testBean.cars}" scrollable="true" scrollWidth="600px" scrollHeight="150">
        <p:column headerText="Id" footerText="Id" width="20%">
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Brand" footerText="Brand" >
            <h:outputText value="#{car.brand}" />
        </p:column>
        <p:column headerText="Year" footerText="Year" >
            <h:outputText value="#{car.year}" />
        </p:column>
        <p:column headerText="Color" footerText="Color" width="30%">
            <h:outputText value="#{car.color}" />
        </p:column>
    </p:dataTable>

		   <h3>Horizontal and Vertical</h3>
    <p:dataTable var="car" value="#{testBean.cars}" scrollable="true" scrollWidth="50%" scrollHeight="150">
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Id" footerText="Id" >
            <h:outputText value="#{car.id}" />
        </p:column>
    </p:dataTable>

    </h:body>

	</f:view>

</html>
`

## 6) Sample bean

use the example bean https://www.primefaces.org/showcase/ui/data/datatable/scroll.xhtml

### PR DESCRIPTION
**Pull request by fanste**
if box-sizing is border-box

### CODE REVIEW

File: src/main/resources/META-INF/resources/primefaces/datatable/datatable.js [hunk 1]
```diff
@@ -1560,14 +1560,46 @@ PrimeFaces.widget.DataTable = PrimeFaces.widget.DeferredWidget.extend({
...
+        }
+
+        if(!width) {
+            width = col.width();
+            isOuterWidth = false;
+        }
+
+        return {
+        	width: width,
+            isOuterWidth: isOuterWidth
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces/datatable/datatable.js:1626]: indenting issue

File: src/main/resources/META-INF/resources/primefaces/datatable/datatable.js [hunk 2]
```diff
@@ -1560,14 +1560,46 @@ PrimeFaces.widget.DataTable = PrimeFaces.widget.DeferredWidget.extend({
...
+        };
+    },
+
+    applyWidthInfo: function(element, widthInfo) {
+        if(widthInfo.isOuterWidth) {
+	        element.outerWidth(widthInfo.width);
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces/datatable/datatable.js]: indenting issue

**Review by melloware — CHANGES_REQUESTED** [review]

File: src/main/resources/META-INF/resources/primefaces/datatable/datatable.js (hunk 2 above)
Comment (fanste) [reply src/main/resources/META-INF/resources/primefaces/datatable/datatable.js]: I hate my eclipse instance...

**Review by melloware — APPROVED** [review]
this looks good to me

File: src/main/resources/META-INF/resources/primefaces/datatable/datatable.js [hunk 3]
```diff
@@ -1560,14 +1560,46 @@ PrimeFaces.widget.DataTable = PrimeFaces.widget.DeferredWidget.extend({
...
-            var diff = element.outerWidth() - element.width();
-            element.width(parseFloat(width) - diff);
+            element.outerWidth(width);
         }
         else {
             element.width(width);
         }
     },
 
+    getColumnWidthInfo: function(col, isIncludeResizeableState) {
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces/datatable/datatable.js:1606]: Can you add JSDOC to this method and test that the build works with `mvn clean package -Pjsdoc` please?

File: src/main/resources/META-INF/resources/primefaces/datatable/datatable.js [hunk 4]
```diff
@@ -1560,14 +1560,46 @@ PrimeFaces.widget.DataTable = PrimeFaces.widget.DeferredWidget.extend({
...
+            isOuterWidth = false;
+        }
+
+        return {
+            width: width,
+            isOuterWidth: isOuterWidth
+        };
+    },
+
+    applyWidthInfo: function(element, widthInfo) {
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces/datatable/datatable.js:1636]: JSDoc

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by christophs78** [comment]
Added a integration-test for sticky-header: https://github.com/primefaces-extensions/primefaces-integration-tests/pull/44
We may add a few more mutations. :-)

File: src/main/resources/META-INF/resources/primefaces/datatable/datatable.js (hunk 3 above)
Comment (fanste) [reply src/main/resources/META-INF/resources/primefaces/datatable/datatable.js:1606]: Done

**Review by melloware — APPROVED** [review]

**Comment by tandraschko** [comment]
i would like to have this reviewed by PrimeTek / @mertsincan

**Comment by melloware** [comment]
@fanste this has some minor conflicts that need to be resolved.

**Comment by fanste** [comment]
Thank you for taking this on. I was on vacation and can only develop for PF on my business computer.

**Comment by melloware** [comment]
No problem. Hopefully Mert will review this week and merge before 10.RC1.

**Comment by mertsincan** [comment]
Hi all,
@fanste Thank alot for your contribution. I'll examine this PR in detail after RC1.

Thanks a lot for your understanding ;)

**Pull request merged**